Add Epic Issues API

justinrush opened the pull request:
This PR adds the methods from the [Epic Issues API](https://docs.gitlab.com/ee/api/epic_issues.html):
* ListEpicIssues
* AssignEpicIssue
* RemoveEpicIssue
* UpdateEpicIssueAssignment

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+	Issue *Issue `json:"issue"`
+}
+
+// AssignEpicIssue assigns an existing issue to an Epic
```
In Go comments should be treated as sentences: https://blog.golang.org/godoc

```suggestion
// AssignEpicIssue assigns an existing issue to an Epic.
```

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+	if err != nil {
+		return nil, nil, err
+	}
+
```
```suggestion
```

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+	}
+
+	var a *EpicIssueAssignment
+
```
```suggestion
```

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+	return a, resp, err
+}
+
+// RemoveEpicIssue removes an issue from an Epic
```
```suggestion
// RemoveEpicIssue removes an issue from an Epic.
```

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+
+// RemoveEpicIssue removes an issue from an Epic
+//
+// Gitlab API Docs: https://docs.gitlab.com/ee/api/epic_issues.html#remove-an-issue-from-the-epic
```
```suggestion
// Gitlab API Docs:
// https://docs.gitlab.com/ee/api/epic_issues.html#remove-an-issue-from-the-epic
```

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+	if err != nil {
+		return nil, nil, err
+	}
+
```
```suggestion
```

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+	}
+
+	var a *EpicIssueAssignment
+
```
```suggestion
```

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+	return a, resp, err
+}
+
+// UpdateEpicIsssueAssignmentOptions describes options to move issues within an epic
```
Same thing about the comment and doc reference.

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+type UpdateEpicIsssueAssignmentOptions struct {
+	*ListOptions
+	MoveBeforeID int `json:"move_before_id"`
+	MoveAfterID  int `json:"move_after_id"`
```
The last two fields should both be `*int` and the tags need to include tags for `url` and both need to contain `omitemtpy`.

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+	MoveAfterID  int `json:"move_after_id"`
+}
+
+// UpdateEpicIssueAssignment moves an issue before or after another issue in an epic issue list
```
Same about the comment...

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+
+// UpdateEpicIssueAssignment moves an issue before or after another issue in an epic issue list
+//
+// Gitlab API Docs: https://docs.gitlab.com/ee/api/epic_issues.html#update-epic---issue-association
```
```suggestion
// Gitlab API Docs:
// https://docs.gitlab.com/ee/api/epic_issues.html#update-epic---issue-association
```

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+	if err != nil {
+		return nil, nil, err
+	}
+
```
```suggestion
```

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+// AssignEpicIssue assigns an existing issue to an Epic
+//
+// Gitlab API Docs: https://docs.gitlab.com/ee/api/epic_issues.html#assign-an-issue-to-the-epic
+func (s *EpicsService) AssignEpicIssue(gid interface{}, epic int, issue int, options ...RequestOptionFunc) (*EpicIssueAssignment, *Response, error) {
```
When declaring two params of the same type, you can write that a bit nicer/shorter:
```suggestion
func (s *EpicsService) AssignEpicIssue(gid interface{}, epic, issue int, options ...RequestOptionFunc) (*EpicIssueAssignment, *Response, error) {
```

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+	}
+
+	var i []*Issue
+
```
```suggestion
```

svanharmelen commented on epic_issues.go:
```diff
@@ -0,0 +1,121 @@
+		return nil, nil, err
+	}
+
+	var i []*Issue
```
Same thing about the name (maybe add an `s`).

justinrush commented on issues.go:
```diff
@@ -111,6 +111,7 @@ type Issue struct {
 	Links                *IssueLinks      `json:"_links"`
 	IssueLinkID          int              `json:"issue_link_id"`
 	MergeRequestCount    int              `json:"merge_requests_count"`
+	EpicIssueID          int              `json:"epic_issue_id"`
```
I wasn't really sure how to handle this.  apparently when issues are returned via the ListEpicIssues API it returns an array of issues that have the epic_issue_id attribute on them.  That id is needed to remove or update the association.

I thought about just creating a new EpicIssue struct and inheriting the Issue struct, but there didn't really seem to be a downside including it in the main Issue struct and we don't introduce another object type.  I can go either way - what's your preference?

svanharmelen replied:
With this context I’m fine with leaving this as is 👍🏻

justinrush commented:
@svanharmelen thanks for patiently dealing with my style issues - still new to golang and i rely on the linters pretty heavily and sometimes they stop working on me.  I think I addressed all your feedback - let me know if there is anything else.
